### PR TITLE
Fix OAuth authorization expiring

### LINKED ISSUE #244: Do I have to manually regenerate the OAuth2 credentials file every few days with GMail?
I've started using yagmail a while ago, and after being very happy at the beginning I discovered something that seems wrong.

After a few days of sending emails, Gmail starts to refuse to send them with the following as part of the stack trace:
```
eturn get_oauth_string(user, oauth2_info)                            
File "/usr/local/lib/p│ython3.9/site-packages/yagmail/oauth2.py", line 96, in get_oauth_string
  access_token, expires_in = refresh_authorization(**oauth2_info)
File "/usr/local/lib/python3.9/site-packages/yagmail/oauth2.py", line 91, in refresh_authorization
  response = call_refresh_token(google_client_id, google_client_secret, google_refresh_token)
File "/usr/local/lib/python3.9/site-packages/yagmail/oauth2.py", line 71, in call_refresh_token
 response = urlopen(request_url, encoded_params).read().decode('UTF-8')
File "/usr/local/lib/python3.9/urllib/request.py", line 214, in urlopen 
  return opener.open(url, data, timeout)                                
File "/usr/local/lib/python3.9/urllib/request.py", line 523, in open    
  response = meth(req, response)                                        
File "/usr/local/lib/python3.9/urllib/request.py", line 632, in http_response
  response = self.parent.error(
File "/usr/local/lib/python3.9/urllib/request.py", line 561, in error   
  return self._call_chain(*args)
File "/usr/local/lib/python3.9/urllib/request.py", line 494, in _call_chain
  result = func(*args)│Ju
  File "/usr/local/lib/python3.9/urllib/request.py", line 641, in http_error_default
 raise HTTPError(req.full_url, code, msg, hdrs, fp)                    
urllib.error.HTTPError: HTTP Error 400: Bad Request                       
```
I've solved it once in the past by regenerating the `credentials.json` locally on my machine and uploading to the server. But now i stopped working again and I'm wondering: am I doing something that's obviously wrong? 

The code that I use is straightforward:
```
    def send_with_yagmail(self):
        yag = yagmail.SMTP(self.our_email, oauth2_file="/<path_to>/credentials.json")
        yag.send(self.to_email, self.message_subject, contents=self.message_body)
```

and the contents of the credentials.json file also seem standard:

```
{
  "email_address": "my.team@gmail.com",
  "google_client_id": "74<blabla>0e1feov.apps.googleusercontent.com",
  "google_client_secret": "GOCSP<blabla<kOz",
  "google_refresh_token": "1//09<blabla>-2_dxZH8"
}

```

Can anybody recommend something that can be done? Thanks!

### LINKED ISSUE #247: Oauth2 refresh token expired after 7 days
Hi, is there any way to extend the refresh token expiration date or automate the process of getting a new refresh token?

### PR DESCRIPTION
**Pull request by xnumad**
Google OAuth authorizations expire after 7 days if issued while the project was in "Testing" mode. [1]

Maybe *the authorization itself* expires, i.e. every token associated with it. So you can't just fix it by refreshing earlier (using the refresh token earlier to get a new auth token). [2] It would anyway be a more complicated setup since refreshing would need to be scheduled.

=> (Google Cloud Console) projects shall use "In Production" status
If your project is in status “in production”, the refresh token will last longer (indefinitely, as I understand). [3]

---

Yagmail currently uses OOB auth which will not be supported in projects with "In Production" status.
=> Switch to a localhost redirect_uri. (which is *still allowed for Desktop type clients*)
[4]

---

Fixes #244, fixes #247

---

[1] https://stackoverflow.com/a/68776672
[2] https://stackoverflow.com/a/67966982
[3] https://stackoverflow.com/questions/8953983/do-google-refresh-tokens-expire
[4] https://developers.google.com/identity/protocols/oauth2/resources/oob-migration#desktop-client

### CODE REVIEW

**Comment by xnumad** [comment]
Although I am certain, **I have not yet confirmed that it works as expected** (i.e. the token won't expire after 7 days) (period not over yet).
Will write again once 7 days are over.

File: yagmail/oauth2.py [hunk 1]
```diff
@@ -99,6 +101,8 @@ def get_oauth_string(user, oauth2_info):
 
 
 def get_oauth2_info(oauth2_file):
+    oauth_setup_readme_link = "See readme for proper setup, preventing authorization from expiring after 7 days!"
```
Comment (kootenpv) [inline yagmail/oauth2.py:104]: We could add a link to readme anchor

**Comment by kootenpv** [comment]
Thanks so much for diving into this!

Code lgtm, let me know if it solves your problem after the 7 days then we can merge and update yagmail on pypi

**Comment by xnumad** [comment]
Still works after 7 days as expected! :+1:

**Comment by kootenpv** [comment]
Awesome thanks for checking

**Comment by kootenpv** [comment]
You can find the new version on pypi:

    pip install -U --no-cache yagmail>=0.15.293

**Comment by yekibud** [comment]
@xnumad Thanks for this feature!  Do you know if there are any security concerns around changing to "In Production" status?  According to Google

> Your app will be available to anyone with a Google account

But, IIUC, this would only happen if you exposed an oauth flow with your client credentials.

image

**Comment by kootenpv** [comment]
I have never seen this option so probably somewhere you selected a wrong option...

**Comment by yekibud** [comment]
This is the only way to convert an oauth app to production so you refresh token doesn't expire.  The message is under the "In Production" tooltip, so you maybe you didn't mouse over it.

image

**Comment by xnumad** [comment]
tl;dr: 👍 This is no problem.

---

> But, IIUC, this would only happen if you exposed an oauth flow with your client credentials.

Yes, that's mandatory anyway (there's no way around it). There's no such thing like a public listing of Google Cloud projects so that your project would be listed in there.

---
"In production" status:
> Your app will be available to anyone with a Google account

I think it's clearer when reading this in contrast to the description for "testing" publishing status:

> Your app will only be available to users you add to the list of test users.

[Documentation for more details](https://support.google.com/cloud/answer/10311615)

So in testing mode, you have to explicitly allow people (Google accounts) before they can authorize your application.

This should not matter for yagmail when the Google account you are authorizing yagmail to is also the one that owns the Google Cloud project (i.e. authorizing to your own app).

**Comment by yekibud** [comment]
Makes sense @xnumad !  Thanks for the peace of mind.  And, once again, thanks for the feature.  Wound up switching to app password credentials, because I couldn't determine whether the oauth feature was really more "secure" via this method of fetching a permanent refresh token.  But if someone with better security chops than me comes along, it's good to know we can switch back to oauth with yagmail.
